libraries/AP_BattMonitor/AP_BattMonitor_ACS37800.cpp

```diff
@@ -0,0 +1,187 @@
+    // @User: Advanced
+    AP_GROUPINFO("C_FINE_M", 28, AP_BattMonitor_ACS37800, curr_fine_mult, 1.0),
+
+    AP_GROUPEND
+};
+
+AP_BattMonitor_ACS37800::AP_BattMonitor_ACS37800(AP_BattMonitor &mon, 
+                            AP_BattMonitor::BattMonitor_State &mon_state, 
+                            AP_BattMonitor_Params &params):
+        AP_BattMonitor_Backend(mon, mon_state, params)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _state.var_info = var_info;
+}
+
+void AP_BattMonitor_ACS37800::init(void)
+{
+    _dev = hal.i2c_mgr->get_device(i2c_bus, i2c_address, 100000, false, 20);
+    if (!_dev) {
+        return;
+    }else {
+        ::printf("ACS37800-> Found device!\n");
+
+        _ctr = 0;
+        WITH_SEMAPHORE(_dev->get_semaphore());
+        // Configure ACS for DC operation 
+        if(write_word(REG_ACCESS_CODE, CUSTOMER_CODE)){
+            ::printf("ACS37800-> Wrote access code successful\n");
+        }
+        hal.scheduler->delay(100);
+        write_word(REG_F_SHADOW, CONFIG_F);
+        uint32_t c_config;
+        read_word(REG_C_SHADOW, c_config);
+        ::printf("ACS37800-> REG_SHADOW_C(read): %x\n", c_config);
+        c_config |= CONFIG_C_MASK;
+        write_word(REG_C_SHADOW, c_config);
+        ::printf("ACS37800-> REG_SHADOW_C(wrote): %x\n", c_config);
+        hal.scheduler->delay(100);
+        // write_word(REG_F_EEPROM, CONFIG_F);
+        // hal.scheduler->delay(100);
+        uint32_t check_c_config;
+        read_word(REG_C_SHADOW,check_c_config);
+        ::printf("ACS37800-> REG_SHADOW_C(read after write): %x\n",check_c_config);
+
+        _dev->register_periodic_callback(25000, FUNCTOR_BIND_MEMBER(&AP_BattMonitor_ACS37800::update, void));
+    }
+}
+
+void AP_BattMonitor_ACS37800::read()
+{
+    WITH_SEMAPHORE(accumulate.sem);
+    _state.healthy = accumulate.count > 0;
+    if (!_state.healthy) {
+        return;
+    }
+
+    _state.voltage = accumulate.volt_sum / accumulate.count;
+    _state.current_amps = accumulate.current_sum / accumulate.count;
+    accumulate.volt_sum = 0;
+    accumulate.current_sum = 0;
+    accumulate.count = 0;
+
+    const uint32_t tnow = AP_HAL::micros();
+    const uint32_t dt_us = tnow - _state.last_time_micros;
+    
+    // update total current drawn since startup
+    update_consumed(_state, dt_us);
+    _state.last_time_micros = tnow;
+}
+
+void AP_BattMonitor_ACS37800::update()
+{
+    WITH_SEMAPHORE(_dev->get_semaphore());
+    acs378002a_t immediate;
+    if (read_word(REG_VCODES_ICODES, immediate.data.all)) {     
+        WITH_SEMAPHORE(accumulate.sem);
+        accumulate.volt_sum += convert_voltage(immediate.data.bits.vcodes);
+        accumulate.current_sum += convert_current(immediate.data.bits.icodes);
+        accumulate.count++;
+    }
+}
+
+float AP_BattMonitor_ACS37800::convert_voltage(int16_t vcode)
+{
+    return (((((float) vcode)/27500.0)*250)/1000)*RESIST_MULT*volt_fine_mult;
```

Comment:
I wonder how well this is optimized by the compiler.

Given this is all multiplicative, could we break out a `constexpr float scalar = ....;` to ensure it is optimised?

similarly elsewhere